src/meili/map_matcher.cc

```diff
@@ -417,19 +422,51 @@ MapMatcher::OfflineMatch(const std::vector<Measurement>& measurements, uint32_t
       const auto& interpolated_results = InterpolateMeasurements(*this, it->second, this_stateid, next_stateid);
 
       // Copy the interpolated match results into the final set
-      std::copy(interpolated_results.cbegin(), interpolated_results.cend(), std::back_inserter(best_paths.back()));
+      std::copy(interpolated_results.cbegin(), interpolated_results.cend(), std::back_inserter(best_path));
+    }
+
+    // TODO: do something to get real cost later, if we dont like sending back cost from here we can simply send
+    // back k as the float value so that at least we know the smaller ones are relatively better than the larger
+    // ones we just dont know how much and then if we want to call back in with the result to get the real cost
+    // we can do that later
+
+    // Keep the path and the cost for it, if you cant find a valid state id we just give it a huge cost
+    auto found_state = std::find_if(stateids.rbegin(), stateids.rend(), [](const StateId& si) {
+      return si.IsValid();
+    });
+    auto accumulated_cost = found_state != stateids.rend() ? vs_.AccumulatedCost(*found_state) : MAX_ACCUMULATED_COST;
+
+    // If we didnt find any valid state we are out of results and should stop
+    //TODO: this doesnt seem to be working
+    if(!best_paths.empty() && found_state == stateids.rend())
+      break;
+
+    // Construct a result
+    auto segments = ConstructRoute(*this, best_path.cbegin(), best_path.cend());
+    MatchResults match_results(std::move(best_path), std::move(segments));
+    match_results.score = accumulated_cost;
+
+    // Keep this path if its unique
+    auto hash = hasher(match_results);
```

Comment:
Is it possible to move this part code to a function `IsSamePath(vector<MatchResult>& rhs, vector<MatchResult>& rhs)` to reduce the complexity of OfflineMatch?

Reply:
sure thing!

Reply:
Also, I assume that if the current path is different from the previous best path, then it should be different from all previous best paths. With this assumption, we don't need to compare it with all previous best paths. WDYT?

Reply:
are you saying we only need to compare to the previous best path and not keep all the hashes? i had the same thought at first but then i was worried that perhaps you could have a simple case like in the issue this references, where its just a simple fork in the road and the first result takes the northern edge of the fork, the second result takes the lower edge of the fork and then the 3rd result node snaps and takes the top edge of the fork again.

also although the code is somewhat complicated in the end im only storing a hash and an index, so there is very little overhead. its basically no extra memory or computation to have all of the paths' hashes to check against. i think the extra protection is worth it because i dont think i can prove that we dont need to account of it picking the same result that we found earlier than the last result. i'm making more test cases and we'll see if maybe we can prove this could be simplified.